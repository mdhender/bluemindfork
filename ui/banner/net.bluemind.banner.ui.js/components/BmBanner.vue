<template>
    <bm-navbar type="dark" variant="info-dark" toggleable="lg" class="bm-banner p-0 align-items-stretch">
        <div id="all-apps-popover" class="px-3 align-self-center">
            <bm-icon icon="9dots" size="2x" />
        </div>
        <bm-popover
            ref="apps-popover"
            target="all-apps-popover"
            placement="bottomright"
            custom-class="apps-popover"
            triggers="click blur"
            variant="info-dark"
            @shown="setFocus"
        >
            <div class="text-white mt-1 mb-2 ml-2">{{ $t("banner.main.apps") }} </div>
            <bm-row class="bm-apps">
                <bm-col v-for="app in applications" :key="app.href" cols="6" class="text-white">
                    <a v-if="app.external" :href="app.href">
                        <div class="pl-3 my-2 bm-app">
                            <bm-banner-app-icon :icon-app="app.icon" />
                            <span class="pl-2 text-uppercase">{{ app.name }}</span>
                        </div>
                    </a>
                    <router-link v-else :to="app.href" tag="div" class="pl-3 my-2 bm-app" @click.native="closePopover">
                        <bm-banner-app-icon :icon-app="app.icon" />
                        <span class="pl-2 text-uppercase">{{ app.name }}</span>
                    </router-link>
                </bm-col>
            </bm-row>
        </bm-popover>
        <bm-navbar-brand :title="$t('banner.main.brand', software)" href="#" :to="logoLink">
            <svg height="30" width="120" v-html="logo" />
        </bm-navbar-brand>
        <bm-banner-applications :applications="defaultApps" />
        <div class="app-title text-center flex-grow-1 align-self-center font-weight-bold text-uppercase">
            <span v-if="selectedApp">{{ selectedApp.name }}</span>
        </div>
        <component :is="widget.component" v-for="widget in widgets" :key="widget.component" />
        <bm-banner-user :user="user" />
    </bm-navbar>
</template>

<script>
import BannerL10N from "@bluemind/banner.l10n";
import BmBannerApplications from "./BmBannerApplications";
import BmBannerAppIcon from "./BmBannerAppIcon";
import BmBannerUser from "./BmBannerUser";
import { BmCol, BmIcon, BmLogo, BmNavbar, BmNavbarBrand, BmPopover, BmRow } from "@bluemind/styleguide";

export default {
    name: "BmBanner",
    components: {
        BmBannerApplications,
        BmBannerAppIcon,
        BmBannerUser,
        BmCol,
        BmIcon,
        BmNavbar,
        BmNavbarBrand,
        BmPopover,
        BmRow
    },
    componentI18N: { messages: BannerL10N },
    props: {
        applications: {
            required: true,
            type: Array
        },
        widgets: {
            required: true,
            type: Array
        },
        software: {
            type: Object,
            required: true
        },
        user: {
            required: true,
            type: Object
        }
    },
    data() {
        return {
            logo: BmLogo,
            selectedAppPath: ""
        };
    },
    computed: {
        selectedApp() {
            return this.applications.find(application => this.$route.path.startsWith(application.href));
        },
        logoLink() {
            return this.selectedApp ? this.selectedApp.href : "/";
        },
        defaultApps() {
            return this.applications.filter(app => 
                app.href === "/mail/" || app.href === "/contact/" || app.href === "/cal/"
            );
        }
    },
    methods: {
        closePopover() {
            this.$root.$emit('bv::hide::popover', 'all-apps-popover');
        },
        setFocus() {
            this.$nextTick(() => document.getElementsByClassName("apps-popover")[0].focus());
        }
    }
};
</script>

<style lang="scss">
@import '~@bluemind/styleguide/css/_variables';

.bm-banner .app-title {
    color: $primary;
}

.bm-banner .fa-9dots, .apps-popover .bm-app {
    cursor: pointer;
}

.bm-banner .fa-9dots {
    color: color-yiq(theme-color("info-dark")) !important;   
}

.bm-banner .navbar-brand {
    margin-right: 0;
}

.apps-popover {
    max-width: unset !important;
    margin-top: map-get($spacers, 3) !important;
    left: -5px !important;
}

.apps-popover .arrow {
    left: 17px !important;
}

.apps-popover .bm-apps {
    width: 22rem;
}

.apps-popover .bm-app {
    border-left: transparent solid 3px;
}

.apps-popover .bm-app:hover, .apps-popover .bm-app.router-link-exact-active {
    color: $primary;
    border-left: $primary solid 3px;
}

.apps-popover a, .apps-popover a:hover {
    all: unset;
}

.apps-popover .popover-body {
    padding-right: 0;
    padding-left: 0;
}
</style>