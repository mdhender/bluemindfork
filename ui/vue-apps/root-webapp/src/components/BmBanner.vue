<template>
    <bm-navbar type="dark" variant="info-dark" class="bm-banner p-0 align-items-stretch d-none d-lg-flex">
        <bm-button
            id="all-apps-popover"
            v-bm-tooltip
            variant="link"
            class="px-3 align-self-center"
            :title="$t('banner.reach.all_apps')"
            :aria-label="$t('banner.reach.all_apps')"
        >
            <bm-icon icon="9dots" size="2x" />
        </bm-button>
        <bm-popover
            ref="apps-popover"
            target="all-apps-popover"
            placement="bottomright"
            custom-class="apps-popover"
            triggers="click blur"
            variant="info-dark"
            @shown="setFocus"
        >
            <div class="text-white mb-2 mx-3">{{ $t("banner.main.apps") }}</div>
            <bm-row class="bm-apps">
                <bm-col v-for="app in applications" :key="app.href" cols="6" class="text-white">
                    <a v-if="app.external" :href="app.href">
                        <div class="pl-3 my-2 bm-app">
                            <bm-banner-app-icon :icon-app="app.icon" />
                            <span class="pl-2 text-uppercase align-middle">{{ app.name }}</span>
                        </div>
                    </a>
                    <router-link v-else :to="app.href" tag="div" class="pl-3 my-2 bm-app" @click.native="closePopover">
                        <bm-banner-app-icon :icon-app="app.icon" />
                        <span class="pl-2 text-uppercase align-middle">{{ app.name }}</span>
                    </router-link>
                </bm-col>
            </bm-row>
        </bm-popover>
        <bm-navbar-brand v-bm-tooltip href="#" :to="logoLink" :title="$t('banner.main.brand')">
            <!-- eslint-disable-next-line vue/no-v-html -->
            <svg height="30" width="120" v-html="logo" />
        </bm-navbar-brand>
        <bm-banner-applications :applications="defaultApps" />
        <div class="app-title text-center flex-grow-1 align-self-center font-weight-bold text-uppercase">
            <span v-if="selectedApp">{{ selectedApp.name }}</span>
        </div>
        <component :is="widget.component" v-for="widget in widgets" :key="widget.component" />
        <bm-banner-user :user="user" @openPreferences="$emit('openPreferences')" />
    </bm-navbar>
</template>

<script>
import BannerL10N from "../../l10n/banner/";
import BmBannerApplications from "./BmBannerApplications";
import BmBannerAppIcon from "./BmBannerAppIcon";
import BmBannerUser from "./BmBannerUser";
import {
    BmButton,
    BmCol,
    BmIcon,
    BmLogo,
    BmNavbar,
    BmNavbarBrand,
    BmPopover,
    BmRow,
    BmTooltip
} from "@bluemind/styleguide";

export default {
    name: "BmBanner",
    components: {
        BmBannerApplications,
        BmBannerAppIcon,
        BmBannerUser,
        BmButton,
        BmCol,
        BmIcon,
        BmNavbar,
        BmNavbarBrand,
        BmPopover,
        BmRow
    },
    directives: { BmTooltip },
    componentI18N: { messages: BannerL10N },
    props: {
        applications: {
            required: true,
            type: Array
        },
        widgets: {
            required: true,
            type: Array
        },
        software: {
            type: Object,
            required: true
        },
        user: {
            required: true,
            type: Object
        }
    },
    data() {
        return {
            logo: BmLogo,
            selectedAppPath: ""
        };
    },
    computed: {
        selectedApp() {
            return this.applications.find(application => this.$route.path.startsWith(application.href));
        },
        logoLink() {
            return this.selectedApp ? this.selectedApp.href : "/";
        },
        defaultApps() {
            return this.applications.filter(
                app => app.href === "/mail/" || app.href === "/contact/" || app.href === "/cal/"
            );
        }
    },
    methods: {
        closePopover() {
            this.$root.$emit("bv::hide::popover", "all-apps-popover");
        },
        setFocus() {
            this.$nextTick(() => document.getElementsByClassName("apps-popover")[0].focus());
        }
    }
};
</script>

<style lang="scss">
@import "~@bluemind/styleguide/css/_variables";

.bm-banner .app-title {
    color: $primary;
}

.bm-banner .fa-9dots,
.apps-popover .bm-app {
    cursor: pointer;
}

.bm-banner .fa-9dots {
    color: color-yiq(theme-color("info-dark")) !important;
}

.bm-banner .navbar-brand {
    margin-right: 0;
}

.bm-banner #all-apps-popover:focus {
    outline: 1px dotted $white;
}

.bm-banner #all-apps-popover:focus,
.bm-banner #all-apps-popover:hover {
    background-color: unset;
    &::before {
        opacity: 0;
    }
}

.apps-popover {
    max-width: unset !important;
    left: -23px !important;
}

.apps-popover .arrow {
    left: 17px !important;
}

.apps-popover .bm-apps {
    width: 22rem;
    a:focus {
        outline: 1px dotted $light;
    }
}

.apps-popover .bm-app:hover,
.apps-popover a:focus,
.apps-popover .bm-app.router-link-active {
    color: $primary;
}

.apps-popover .bm-app.router-link-active {
    font-weight: $font-weight-bold;
}

.apps-popover a,
.apps-popover a:hover {
    all: unset;
}

.apps-popover .popover-body {
    padding-right: 0;
    padding-left: 0;
}
</style>
