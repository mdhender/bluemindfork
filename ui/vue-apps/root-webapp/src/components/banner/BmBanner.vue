<template>
    <bm-navbar :aria-label="$t('banner.label')" variant="primary" class="bm-banner align-items-center d-none d-lg-flex">
        <bm-icon-button
            id="all-apps-popover"
            variant="regular-on-fill-primary"
            size="lg"
            icon="9dots"
            :title="$t('banner.reach.all_apps')"
            :aria-label="$t('banner.reach.all_apps')"
        />
        <bm-banner-applications :applications="applications" target="all-apps-popover" />
        <bm-navbar-brand href="#" :to="logoLink" :title="$t('banner.main.brand')">
            <img src="images/logo-bluemind.png" alt="" />
        </bm-navbar-brand>
        <bm-banner-shortcuts :applications="defaultApps" />
        <div class="app-title text-center flex-grow-1 font-weight-bold text-uppercase">
            <span v-if="selectedApp">{{ selectedApp.name }}</span>
        </div>
        <bm-extension id="webapp.banner" path="widget" class="d-flex align-items-center" />
        <bm-banner-help v-if="selectedApp && selectedApp.help" :url="selectedApp.help" />
        <bm-banner-menu :user="user" />
    </bm-navbar>
</template>

<script>
import { BmIconButton, BmNavbar, BmNavbarBrand } from "@bluemind/styleguide";
import { BmExtension } from "@bluemind/extensions.vue";
import BannerL10N from "../../../l10n/banner/";
import BmBannerApplications from "./BmBannerApplications";
import BmBannerMenu from "./BmBannerMenu";
import BmBannerHelp from "./BmBannerHelp";
import BmBannerShortcuts from "./BmBannerShortcuts";

export default {
    name: "BmBanner",
    components: {
        BmBannerApplications,
        BmBannerHelp,
        BmBannerMenu,
        BmBannerShortcuts,
        BmExtension,
        BmIconButton,
        BmNavbar,
        BmNavbarBrand
    },
    componentI18N: { messages: BannerL10N },
    props: {
        applications: {
            required: true,
            type: Array
        },
        user: {
            required: true,
            type: Object
        }
    },
    computed: {
        selectedApp() {
            return this.applications.find(application => this.$route.path.startsWith(application.path));
        },
        logoLink() {
            return this.selectedApp ? this.selectedApp.path : "/";
        },
        defaultApps() {
            return this.applications.filter(({ shortcut }) => shortcut);
        }
    }
};
</script>

<style lang="scss">
@import "~@bluemind/styleguide/css/_variables";

.bm-banner {
    &.navbar {
        flex: 0 0 42px;

        padding: 0;
        padding-left: $sp-4;
    }

    #all-apps-popover:focus,
    #all-apps-popover:hover {
        &::before {
            opacity: 0;
        }
    }

    .fa-9dots {
        cursor: pointer;
        margin-right: $sp-3;
    }

    .navbar-brand {
        margin-right: $sp-5;
        padding: 0;
        display: inline-flex;
        align-items: center;
        text-align: center;

        img {
            width: 130px;
        }
    }

    .bm-banner-user {
        margin-left: $sp-6;
    }

    .app-title {
        color: $secondary-fg;
    }
}
</style>
