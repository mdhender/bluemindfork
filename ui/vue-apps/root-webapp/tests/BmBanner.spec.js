import { shallowMount, createLocalVue } from "@vue/test-utils";
import merge from "lodash.merge";
import BmBanner from "../src/components/BmBanner";
import { BmPopover } from "@bluemind/styleguide";

jest.mock("@bluemind/styleguide/css/_variables.scss", () => ({ iconsColors: "" }));
//TODO refacto createWrapper later with testutils when FEATWEBML-597 is merged
function createWrapper(overrides) {
    const localVue = createLocalVue();
    const defaultMountingOptions = {
        localVue,
        propsData: { applications: [], user: {} },
        mocks: {
            $t: () => {},
            $tc: () => {}
        }
    };
    const mergedMountingOptions = merge(defaultMountingOptions, overrides);
    return shallowMount(BmBanner, mergedMountingOptions);
}

describe("BmBanner", () => {
    test("is a Vue instance", () => {
        const wrapper = createWrapper();
        expect(wrapper.isVueInstance()).toBeTruthy();
    });

    test("should match snapshot", () => {
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });

    test("should show menu when button is clicked", () => {
        const wrapper = createWrapper();
        expect(!wrapper.find(BmPopover).isVisible()).toBeTruthy;
        wrapper.find("#all-apps-popover").trigger("click");
        expect(wrapper.find(BmPopover).isVisible()).toBeTruthy();
    });
});
