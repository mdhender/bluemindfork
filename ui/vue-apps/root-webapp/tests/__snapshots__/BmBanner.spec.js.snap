// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BmBanner should match snapshot 1`] = `
<bm-navbar-stub
  class="bm-banner p-0 align-items-stretch d-none d-lg-flex"
  tag="nav"
  type="dark"
  variant="info-dark"
>
  <bm-button-stub
    class="px-3 align-self-center"
    event="click"
    id="all-apps-popover"
    routertag="a"
    tag="button"
    target="_self"
    type="button"
    variant="link"
  >
    <bm-icon-stub
      icon="9dots"
      size="2x"
    />
  </bm-button-stub>
   
  <bm-popover-stub
    boundary="scrollParent"
    boundarypadding="5"
    customclass="apps-popover"
    delay="50"
    fallbackplacement="flip"
    offset="0"
    placement="bottomright"
    target="all-apps-popover"
    triggers="click blur"
    variant="info-dark"
  >
    <div
      class="text-white mb-2 mx-3"
    >
      
    </div>
     
    <bm-row-stub
      class="bm-apps"
      tag="div"
    />
  </bm-popover-stub>
   
  <bm-navbar-brand-stub
    event="click"
    href="#"
    routertag="a"
    tag="div"
    target="_self"
    to="/"
  >
    <svg
      height="30"
      width="120"
    />
  </bm-navbar-brand-stub>
   
  <bm-banner-applications-stub
    applications=""
  />
   
  <div
    class="app-title text-center flex-grow-1 align-self-center font-weight-bold text-uppercase"
  >
    <!---->
  </div>
    
  <bm-banner-user-stub
    user="[object Object]"
  />
</bm-navbar-stub>
`;
